### ISSUE TITLE
Linking delay on container startup

### ISSUE THREAD

**Issue by andrefreitas** [rancher/rancher#2015]
There is a delay on the linking. For example, if I launcher service A and then service B that is linked to A, the link is not available immediately.

# Compose

```yaml
flags:
  image: <someprivateregistry>/flags:0.1.4
  labels:
    io.rancher.container.network: true
  links:
    - etcetera
```

# Issue example
```sh
> rancher-compose -p services up flags                                                      master [6f23648] modified
INFO[0000] Project [services]: Starting project
INFO[0000] [0/25] [flags]: Starting
INFO[0000] Creating service flags
INFO[0067] [1/25] [flags]: Started
INFO[0067] Project [services]: Project started
flags_1 | 2015-09-10T10:16:28.795323761Z Get http://etcetera:4021/docker/flags/settings.json: dial tcp: lookup etcetera: no such host
flags_1 | 2015-09-10T10:16:44.917910167Z Get http://etcetera:4021/docker/flags/settings.json: dial tcp: lookup etcetera on 169.254.169.250:53: no such host
flags_1 | 2015-09-10T10:16:47.372933077Z Get http://etcetera:4021/docker/flags/settings.json: dial tcp: lookup etcetera on 169.254.169.250:53: no such host
flags_1 | 2015-09-10T10:17:03.030846754Z Get http://etcetera:4021/docker/flags/settings.json: dial tcp: lookup etcetera on 169.254.169.250:53: no such host
```
# Workaround
My workaround was doing a ping before starting the server in __docker-entrypoint.sh__

```sh
#!/bin/sh

until ping -c 1 etcetera
do
  echo "trying to connect with etcetera"
done

 /bin/serve
```

# Result
Launching the service again with this health check now works:
```sh
> rancher-compose -p services up flags                                                      master [6f23648] modified
INFO[0000] Project [services]: Starting project
INFO[0000] [0/25] [flags]: Starting
INFO[0000] Creating service flags
INFO[0014] [1/25] [flags]: Started
INFO[0014] Project [services]: Project started
flags_1 | 2015-09-10T10:55:59.260509142Z ping: bad address 'etcetera'
flags_1 | 2015-09-10T10:55:59.260745544Z trying to connect with etcetera
flags_1 | 2015-09-10T10:55:59.263403489Z ping: bad address 'etcetera'
flags_1 | 2015-09-10T10:55:59.263671925Z trying to connect with etcetera
flags_1 | 2015-09-10T10:55:59.266400266Z ping: bad address 'etcetera'
flags_1 | 2015-09-10T10:55:59.266563970Z trying to connect with etcetera
flags_1 | 2015-09-10T10:55:59.268141571Z ping: bad address 'etcetera'
flags_1 | 2015-09-10T10:55:59.268309292Z trying to connect with etcetera
flags_1 | 2015-09-10T10:55:59.270777216Z ping: bad address 'etcetera'
flags_1 | 2015-09-10T10:55:59.270944841Z trying to connect with etcetera
flags_1 | 2015-09-10T10:55:59.272289657Z ping: bad address 'etcetera'
flags_1 | 2015-09-10T10:55:59.273013786Z trying to connect with etcetera
flags_1 | 2015-09-10T10:55:59.275089792Z ping: bad address 'etcetera'
flags_1 | 2015-09-10T10:55:59.275260309Z trying to connect with etcetera
flags_1 | 2015-09-10T10:55:59.277277227Z ping: bad address 'etcetera'
flags_1 | 2015-09-10T10:55:59.277427749Z trying to connect with etcetera
flags_1 | 2015-09-10T10:55:59.278871711Z ping: bad address 'etcetera'
flags_1 | 2015-09-10T10:55:59.279050715Z trying to connect with etcetera
flags_1 | 2015-09-10T10:55:59.280443254Z ping: bad address 'etcetera'
flags_1 | 2015-09-10T10:55:59.280587407Z trying to connect with etcetera
flags_1 | 2015-09-10T10:55:59.282159528Z ping: bad address 'etcetera'
flags_1 | 2015-09-10T10:55:59.282320171Z trying to connect with etcetera
flags_1 | 2015-09-10T10:55:59.997826839Z PING etcetera (10.42.46.145): 56 data bytes
flags_1 | 2015-09-10T10:55:59.997894708Z 64 bytes from 10.42.46.145: seq=0 ttl=62 time=714.008 ms
flags_1 | 2015-09-10T10:55:59.997914934Z
flags_1 | 2015-09-10T10:55:59.997923200Z --- etcetera ping statistics ---
flags_1 | 2015-09-10T10:55:59.997930117Z 1 packets transmitted, 1 packets received, 0% packet loss
flags_1 | 2015-09-10T10:55:59.997938171Z round-trip min/avg/max = 714.008/714.008/714.008 ms
flags_1 | 2015-09-10T10:56:00.016096667Z 2015/09/10 10:56:00 Listening at 0.0.0.0:4020	212ns
```

**Comment by ebuildy** [comment]
I believe Rancher has a "post network" step, where agents are setuping DNS, maybe is for this reason.

---------------------

 BTW I found it's always a good practice to wait/check for service status like you do, even if the DNS resolves, most of the time your linked service (for instance elasticsearch) will not be ready as the container starts:

````shell
#!/bin/bash

echo "Waiting elasticSearch up "

until $(curl --output /dev/null --silent --head --fail http://elasticsearch:9200); do
	printf '.'
	sleep 1
done

echo ""

echo "Running Kibana"

/opt/kibana/bin/kibana
````

**Comment by andrefreitas** [comment]
@ebuildy I agree, I am doing this health check in all my containers.

I am wondering if something can be improved in Rancher to prevent this since we are declaring the links in the compose file.

**Comment by ebuildy** [comment]
The solution would be  to determine  IP adresses before create  container
(in  Rancher)
Le 11 sept. 2015 12:25 AM, "André Freitas" <notifications@github.com> a
écrit :

**Comment by andrefreitas** [comment]
@ebuildy having a rancher agent just for DNS can be another alternative? Just wondering.

**Comment by ebuildy** [comment]
I am not sure, I believe the agent is connected via a web socket to Rancher server (aka Cattle), so its the Rancher server that give order to agents.

A (very ugly) solution could be to specify static IPs via labeling (http://docs.rancher.com/rancher/labels/) and hardcode this into your container...

**Comment by deniseschannon** [comment]
This sounds like it would be related to  https://github.com/rancher/rancher/issues/278. Is there always a delay in linking or just the first time?

**Comment by janeczku** [comment]
I have been experiencing this with links between containers on the same host too. E.g. i have a service with a Redis sidekick: The first connection attempt to to the sidekick right after container startup fails every time. Consecutive connections are working.
This makes me think that this is not related to https://github.com/rancher/rancher/issues/278 since there wouldn't be an IPSec tunnel between the two containers.

**Comment by vincent99** [comment]
There is also https://github.com/rancher/rancher/issues/2341 where a container can startup and try to resolve something before DNS is configured.  Or it could be as simple as both containers are started at nearly the same time and the Redis process isn't actually listening yet by the time the client tries to connect to it.

**Comment by janeczku** [comment]
@vincent99 This is a very good point. It hadn't crossed my mind at first since Redis starts up really fast. But it looks like the primary service is even a tad faster. Lesson learned: With sidekicks you have to take care of the startup connection race between sidekick and primary service.

**Comment by deniseschannon** [comment]
@andrefreitas With #278 and #2341, I'm going to close this issue as I believe this is a duplicate of one of those 2 use cases.
